File: datastruct/map/LinkedListMap.java
public class LinkedListMap<K,V> implements Map<K, V> {

    private class Node{
        public K key;
        public V value;
        public Node next;

        public Node(K key, V value, Node next) {
            this.key = key;
            this.next = next;
            this.value = value;
        }

        public Node(K key){
            this(key, null, null);
        }

        public Node(){
            this(null, null, null);
        }

        @Override
        public String toString() {
            return key.toString() + ":" + value.toString();
        }
    }

    private Node dummyHead;
    private int size;

    public LinkedListMap(){
        dummyHead = new Node();
        size = 0;
    }

    @Override
    public int getSize() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    // 辅助方法，获取键为 key 的节点
    private Node getNode(K key){

        Node cur = dummyHead.next;

        while(cur != null){
            if(cur.key.equals(key))
                return cur;
            cur = cur.next;
        }
        return null;
    }

    @Override
    public boolean contain(K key) {
        return getNode(key) != null;
    }

    @Override
    public V get(K key) {
        Node node = getNode(key);

        return node == null ? null: node.value;
    }

    @Override
    public void add(K key, V value) {
        Node cur = getNode(key);

        if(cur == null){
            dummyHead.next = new Node(key, value, dummyHead.next);
            size++;
        }else {
            cur.value = value;
        }
    }

    @Override
    public void set(K key, V value) {
        Node cur = getNode(key);

        if ( cur == null) {
            throw  new IllegalAccessError(key + " doesn't exist");
        }else {
            cur.value = value;
        }
    }

    @Override
    public V remove(K key) {
        Node pre = dummyHead;

        while (pre.next != null) {
            if (pre.next.key.equals(key)) {
                Node cur = pre.next;
                pre.next = cur.next;
                cur.next = null;
                size--;
                return cur.value;
            }
            pre = pre.next;
        }

        return null;
    }
}

File: datastruct/map/LinkedListMap.py
from Map import Map


class Node(object):
    def __init__(self, key=None, value=None, next=None):
        self.key = key
        self.value = value
        self.next = next


class LinkedListMap(Map):

    def __init__(self):
        self.__size = 0
        self.__dummy_head = Node()

    def is_empty(self):
        return self.__size == 0

    def get_size(self):
        return self.__size


    def __get_node(self, key):
        """辅助函数，寻找 key 所在的结点"""
        cur = self.__dummy_head.next

        while cur is not None:
            if cur.key == key:
                return cur

            cur = cur.next

        return None

    def contains(self, key):

        return self.__get_node(key) is not None

    def get(self, key):

        node = self.__get_node(key)

        if node is None:
            return None
        else:
            return node.value

    def add(self, key, value):

        node = self.__get_node(key)

        if node is None:
            self.__dummy_head.next = Node(key, value, self.__dummy_head.next)
            self.__size += 1
        else:
            node.value = value

    def set(self, key, value):

        node = self.__get_node(key)

        if node is None:
            raise IndexError("%s doesn't exist!" % key)
        else:
            node.value = value

    def remove(self, key):

        pre = self.__dummy_head

        while pre.next is not None:
            if pre.next.key == key:
                ret = pre.next
                pre.next = ret.next
                ret.next = None
                self.__size -= 1
                return ret.value

            pre = pre.next

        return None

